### ISSUE TITLE
Create machinekit-cnc stack

### ISSUE THREAD

**Issue by ArcEye** [machinekit/machinekit#1354]
This thread is to record discussions 'off-list' and document progress.

The goal is to separate the machinekit HAL portion of the repo from the CNC portion.

The HAL portion is considerably advanced from the legacy Linuxcnc code and completely incompatible with it.  It is in this area that most of the development effort and expertise is.

This split was made some time back and Machinekit-HAL is stable and fully usable as a single entity.
https://github.com/machinekit/Machinekit-HAL

The CNC portion has diverged little from the Linuxcnc legacy code.
The tp code enhancements were ported across for us a couple of years ago, but aside from emc/rs274/previewmodule.cc which interfaces with machinetalk / czmq, it is largely the same.

Separating the stacks will make working on the CNC side much easier and open up several possibilities including:

1. Not actively developing the CNC stack further and porting enhancements from Linuxcnc when they are judged mature, concentrating on the HAL stack.

2. Replacing the NML messaging system with machinetalk or similar, allowing both stacks to operate
independently to a certain extent, with inter-communication.

3. Replacement of the elements of the CNC stack with 'drop in' modules.
If a common interface is established, elements like the interpreter could be exchanged for another module, perhaps one that works with STEP-NC code for instance.

**Comment by ArcEye** [comment]
Success !

I have stripped out the CNC stack and hacked the makefiles etc.so that the required elements build.

It also passes the acid test, build machinekit-hal then copy across machinekit-cnc binaries and libs into it.  It then runs just as machinekit would.

Because these are RIP builds, to get the scripts etc. to align looking for the same $EMC2_HOME, 
I built both in duplicate dirs of the same name, renaming one whilst the other was built.
ie.
```
# !/bin/bash

git clone https://github.com/machinekit/machinekit-hal.git machinekit-test ;
cd machinekit-test/src;
./autogen.sh && ./configure && make -j$(numproc) && make setuid;
cd ../../
mv machinekit-test machinekit-hal;
git clone https://github.com/arceye/machinekit-cnc.git machinekit-test;
cd machinekit-test/src;
./autogen.sh && ./configure && make -j$(numproc);
cd ../../
mv machinekit-test machinekit-cnc;
mv machinekit-hal machinekit-test;
yes no | cp -riv dir1/* dir2/ ;
```
(the last bit is rather messy but rsync did not work as expected)

The repo is at `https://github.com/arceye/machinekit-cnc` feel free to play :)

**Comment by luminize** [comment]
@ArcEye great! I've added a project on my own repo https://github.com/luminize/machinekit-projects where I'll add all parts and pieces.

**Comment by ArcEye** [comment]
> @luminize I've added a project on my own repo https://github.com/luminize/machinekit-projects where I'll add all parts and pieces.

That looks interesting, look forward to the clear overviews.

**Comment by ArcEye** [comment]
The biggest 'problem' with the `machinekit-cnc` stack, is the dependence upon headers and libs from `machinekit-hal`, which to an extent are moving targets.

There is already requirement for the generated _protobuf headers_ because of the machinetalk code in emc/rs274/previewmodule.cc.
If we replace NML with machinetalk, that will increase.
These could be subject to change at any time and need to be the same ones as in `machinekit-hal` and thus in `machinekit` proper.

Other sections of the code require access to the _hal_ headers and _rtapi_ headers, which likewise are not fixed, even if less likely subject to change than the _pb.h_ and _npb.h_ headers

Most of the libs not actually built as part of the `machinekit-cnc` build are also still required in final linkage.

This makes `machinekit-cnc` not really suitable to be built as a RIP.  
It cannot 'run in place' as a lone entity in any case because of other basic dependencies, 
such as realtime / halcmd etc.

The solution I envisage is to make `machinekit-cnc` a package only.
It can be built in parallel with `machinekit-hal` and headers and libs from that build imported, ensuring
that they are always up to date and match those in `machinekit-hal`.

Realistically, if you are building `machinekit-hal` as a RIP and want the CNC stack, you are just going to build `machinekit` proper in the first place, but the way would still be open to do a 'user install' of the CNC elements and adjust RIP paths to use them.

So that is my next stage, packaging in parallel with the `machinekit-hal` packages.
The extraneous headers and libs will be removed from the `machinekit-cnc` repo, which will render it unable to build except within the CI environment.
Need to comment on that in the README.

I have my own Jenkins systems set up, of which the machinekit ones are copies, so can do it quite simply without lots of PRs etc.

**Comment by sirop** [comment]
@ArcEye 
At the risk of hijacking the theme:
Is there any hope that machinekit-HAL will get GPL-free in the course of all these changes?

**Comment by ArcEye** [comment]
> @sirop 
At the risk of hijacking the theme:
Is there any hope that machinekit-HAL will get GPL-free in the course of all these changes?

You will need to explain what you mean.

Are we going to release with no copyright whatsoever, say anyone can use it and pretend they wrote it too?
Not likely.

**Comment by sirop** [comment]
Well, probably not.
But just to explain what I mean: there are some source files like https://github.com/machinekit/machinekit/blob/master/src/rtapi/rtapi_app.cc which is GPL, however most of the other files are LGPL. There is quite a difference between GPL and LGPL.

**Comment by ArcEye** [comment]
I suggest you open a new Issue, suggesting that all files have a uniform Gnu Public Licence and putting forward your preference and reasoning.
( and yes I did spell `licence` correctly, the US spelling has no concept of noun and verb 😄 )

**Comment by ArcEye** [comment]
Sitrep:
Now have a `machinekit-cnc` package building.

Next the hard work, go through the `machinekit-hal` package against the `machinekit-cnc` one and make sure nothing is duplicated between them.

Currently there are overlaps, some items are quite grey as to whether they are HAL or CNC, because for instance, they may relate to CNC largely, but use hal library calls.

You can get away with murder in a RIP build, but dpkg will not overwrite a file that was installed by a previous package, so the match has to be exact.

If in doubt they will go into `machinekit-hal`, because `machinekit-cnc`  is never going to be installed without it, but the opposite is not true.

**Comment by ArcEye** [comment]
Package `machinekit-cnc`is now building fine, with all HAL or partly HAL elements removed.

A strategy of having /fakelib and /fakeinclude dirs with the required 'out of tree' libs and headers
from `machinekit-hal` is working fine on builds and should allow the required libs/headers to be
added pre-build and ensure they are always up to date.

Next work is to strip out any overlapping elements of `machinekit-hal`, which now reside in the cnc stack.

After that, begin work on a CI builder which uses the `machinekit-hal` build to supply files required for the corresponding `machinekit-cnc` build of the same flavour.

**Comment by ArcEye** [comment]
Finally succeeded creating the machinekit-cnc packages.

Had to create a rather elaborate build sequence which combines the `machinekit-hal` build with the `machinekit-cnc` builds.  
This is probably the most straightforward way to have the correct flavour libs and the correct version headers available for each build.

Packages are available for testing at http://deb.mgware.co.uk

Tested in a bare Jessie partition, installing `machinekit-hal` and then installing `machinekit-cnc`, results in a functioning `machinekit`, able to run axis sims 😄 

Some more polishing to do, then transfer to machinekit repo and create the CI builders there, but the main work is done.

**Comment by luminize** [comment]
Great results @ArcEye !

**Comment by ArcEye** [comment]
Think I am ready to start work on the machinekit org repo.

I am going to change the name of the `Machinekit-HAL` repo to `machinekit-hal`.  
I just followed the 'camel-case' name of the original test dir, but it makes more sense to have all lower case repo names, especially as there cannot be uppercase chars in package names.

I will need to do test runs, so will make innocuous changes and merge them myself to generate rebuilds, then reset the HEAD back to its previous position when all done.

**Comment by ArcEye** [comment]
That proved easier than I dared hope, all the hard work was done in my own Jenkins setup.
No need for test merges, I just did a manual build trigger and it worked first time.

**We now have machinekit-cnc packages**
http://deb.machinekit.io/debian/pool/main/m/machinekit-cnc/

There is also a PR tester for `machinekit-cnc` which I have just tested and that is OK.
For some reason the first run through triggered 2 webhook notifications, the first of which it could not resolve and failed.
Seem to recall it did this when I initiated the other PR builders, but does not do it again thereafter.

I also need to update the `machinekit-hal` README to match the one for `machinekit-cnc`,
but will do a normal PR for that.

All sorted - hope someone makes use of it.

**Comment by ArcEye** [comment]
This really needs testing now.

The particular area of uncertainty for me, is whether any of the python elements still in `machinekit-cnc`, need to be available within `machinekit-hal`.

@machinekoder @luminize @cdsteinkuehler can you have a look and advise?

The `machinekit-hal` package needs to be completely _stand-alone_, it does not matter what dependencies the `machinekit-cnc` package has, as it is incapable of being used without the HAL stack anyway.
( just makes the package build more complicated 😄 )

I recently moved the _latency-test_ stuff across to `machinekit-cnc` because that is where the pyvcp and linuxcnc python libs are, which it uses.
If there is a requirement for pyvcp or gladevcp in the `machinekit-hal` package, that will entail a lot more stuff being moved back over.

The current rationale is based on my usage of it, which is _pyvcp / gladevcp_ is just for panels with Axis or similar, running a CNC machine.

**Comment by luminize** [comment]
> On 16 Feb 2018, at 17:40, ArcEye <notifications@github.com> wrote:
> 
> This really needs testing now.
> 
> The particular area of uncertainty for me, is whether any of the python elements still in machinekit-cnc, need to be available within machinekit-hal.
> 
> @machinekoder @luminize @cdsteinkuehler can you have a look and advise?
> 

Ill have a look this weekend. Ill install both and i'll install machinekit-hal only and give it a spin. Do you have machinekit-cnc packages somewhere or is it best to install machinekit-cnc next to machinekit-hal?
Can one install a machinekit-cnc package along a rip machinekit-hal install?

**Comment by luminize** [comment]
@ArcEye here's my setup:
vanilla debian stretch
stock rt-preempt kernel
rip packages of machinekit and machinekit-hal

So testing out installation of machinekit-cnc as per instructions in the readme

shortened version of apt install method
```
sudo apt install machinekit-cnc-rt-preempt
machinekit-cnc-rt-preempt : Depends: machinekit-hal-rt-preempt but it is not going to be installed
...
sudo apt-get install machinekit-hal-rt-preempt
machinekit-hal-rt-preempt : Depends: python-gst0.10 but it is not installable
...
sudo apt-get install python-gst0.10
Package python-gst0.10 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
E: Package 'python-gst0.10' has no installation candidate
```

**Comment by ArcEye** [comment]
Fixed in https://github.com/machinekit/machinekit-hal/pull/53

**Comment by luminize** [comment]
@ArcEye works! played around with the simulator as of yet, will check more.
great achievement!

**Comment by ArcEye** [comment]
Great.  Thanks for testing.

**Comment by ArcEye** [comment]
No longer relevant - repo exists - closing
